### ISSUE TITLE
pyg_lib, torch_sparse, torch_cluster optional wheels do not work with PyTorch 2.x and cu118

### ISSUE THREAD

**Issue by felker** [pyg-team/pytorch_geometric#8128]
### 🐛 Describe the bug

I build and maintain a series of shared Anaconda installations on an large A100 cluster for the last few years, spanning versions of PyTorch, PyG, and many other packages. 

First, the latest PyTorch is built from source, then PyG and all its optional dependencies are installed from wheels. For the last two installs, that has looked like:
```
pip install pyg_lib torch_sparse torch_cluster torch_scatter torch_spline_conv -f https://data.pyg.org/whl/torch-2.0.1+cu118.html
```

In an environment I built in January 2023 against PyTorch 1.13.1, and the same cuDNN and CUDA, that install line (with changed URL https://data.pyg.org/whl/torch-1.13.1+cu117.html) worked totally fine. 

However, three of the wheels dont seem to work against PyTorch 2.0.1; I get the classic version mismatch error when trying to `import pytorch_geometric`:
```
OSError: /soft/datascience/conda/2023-10-02/mconda3/lib/python3.10/site-packages/torch_cluster/_version_cuda.so: undefined symbol: _ZN5torch3jit17parseSchemaOrNameERKSs
```

After experimenting a bit, I determined that the workaround was to build pyg_lib (0.2.0), torch_sparse (), torch_cluster () from source (the other two optional dep wheels seem fine for now):
```
pip install torch_scatter torch_spline_conv -f https://data.pyg.org/whl/torch-2.0.1+cu118.html                                                                                              # build the rest from source:
pip install --verbose git+https://github.com/pyg-team/pyg-lib.git
pip install --verbose torch_sparse
pip install --verbose torch_cluster
```

I can open issues in each of the 3x repositories, but thought I would report it here first. 

### Environment

* PyG version: 2.3.1
* PyTorch version: 2.0.1
* OS: SUSE Linux Enterprise Server 15 SP3
* Python version: 3.10.12
* CUDA/cuDNN version: 11.8 / 8.6.0.163
* How you installed PyTorch and PyG (`conda`, `pip`, source): Source
* Any other relevant information (*e.g.*, version of `torch-scatter`):

**Comment by rusty1s** [comment]
Yes, that is sadly correct. This is due to the PyTorch ABI version where extensions are only valid if they were built with the same PyTorch version that is being used. For the wheels, we build them against the official PyTorch wheels, and so they work fine whenever you are installing PyTorch as outlined in https://pytorch.org, but break whenever you install PyTorch from source. I am sadly not sure how to fix this :(

**Comment by felker** [comment]
I built PyTorch specifically from e9ebda2 = https://github.com/pytorch/pytorch/tree/v2.0.1, which is the commit that they build their 2.0.1 wheel from. I do apply a single line patch (https://github.com/felker/conda_install_scripts/blob/master/alcf_polaris/hardcode_aten_cudnn.patch) to fix an `#include` directive in their source code.

I am fairly certain that the PyG optional library wheels worked with this approach in the past (but maybe I was always building from source pre-PyG 2.3.x, since there were required libraries)... although maybe for my source build this time there was some small difference like the patch or in the cuDNN, etc. library versions that could result in mismatch with the PyTorch wheel ABI? I can't find much public documentation on the PyTorch ABI or what would affect it, but my general knowledge of C ABI compatibility leads me to believe that it is unlikely. As long as the C++ source of PyTorch and the CUDA toolkit match those used to generate the public wheels, it seems my build should be compatible with the PyG extensions built like this:
https://github.com/rusty1s/pytorch_scatter/blob/master/.github/workflows/building.yml

I wonder if anything can be gleaned from the name mangling e.g. `_ZN5torch3jit17parseSchemaOrNameERKSs`
I assume the ABI check isnt something trivial like checking the version string? If you build PyTorch 2.0.1 from source e.g., it ends up with a version string like `2.0.0a+git8373` unless you set `PYTORCH_BUILD_VERSION`. 

Also, not sure if `_GLIBCXX_USE_CXX11_ABI=1` (which seems to be automatically set in my build) might factor into this since the official PyTorch wheels were built with the opposite convention:
```
>>> import torch
>>> torch._C._GLIBCXX_USE_CXX11_ABI
False
```

**Comment by rusty1s** [comment]
I am not an expert on this either, sorry :( This happens frequently for users who, e.g., install `pytorch` from `-c conda-forge` as well. `_GLIBCXX_USE_CXX1_ABI=0` might be something worth trying out.
